### ISSUE TITLE
Unable to start DynamoDB Local process!

### ISSUE THREAD

**Issue by Alex-RR** [99xt/serverless-dynamodb-local#210]
Any help/idea would be very much appreciated, I'm new to all this serverless thing and want to learn. Thank you in advance!

**Actual Behaviour**

I'm unable to start DynamoDB locally and general error: Unable to start DynamoDB Local process!, I followed the steps in multiple tutorials, none will make it start.

**Expected Behaviour**

DynamoDB should be running smoothly on port 8000

**Steps to reproduce it**

Running the commands "sls dynamodb install" and "sls dynamodb start -p 8000"

**LogCat for the issue**

`Serverless: Load command config
Serverless: Load command config:credentials
Serverless: Load command create
Serverless: Load command install
Serverless: Load command package
Serverless: Load command deploy
Serverless: Load command deploy:function
Serverless: Load command deploy:list
Serverless: Load command deploy:list:functions
Serverless: Load command invoke
Serverless: Load command invoke:local
Serverless: Load command info
Serverless: Load command logs
Serverless: Load command metrics
Serverless: Load command print
Serverless: Load command remove
Serverless: Load command rollback
Serverless: Load command rollback:function
Serverless: Load command slstats
Serverless: Load command plugin
Serverless: Load command plugin
Serverless: Load command plugin:install
Serverless: Load command plugin
Serverless: Load command plugin:uninstall
Serverless: Load command plugin
Serverless: Load command plugin:list
Serverless: Load command plugin
Serverless: Load command plugin:search
Serverless: Load command config
Serverless: Load command config:credentials
Serverless: Load command rollback
Serverless: Load command rollback:function
Serverless: Load command create_domain
Serverless: Load command delete_domain
Serverless: Load command dynamodb
Serverless: Load command dynamodb:migrate
Serverless: Load command dynamodb:seed
Serverless: Load command dynamodb:start
Serverless: Load command dynamodb:noStart
Serverless: Load command dynamodb:remove
Serverless: Load command dynamodb:install
Serverless: Load command offline
Serverless: Load command offline:start
Serverless: Invoke dynamodb:start
 
  Error --------------------------------------------------
 
  Unable to start DynamoDB Local process!
 
     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
 
  Stack Trace --------------------------------------------
 
Error: Unable to start DynamoDB Local process!
    at Object.start (/Users/alexraducu/work/serverless/ssr-react-next/node_modules/dynamodb-localhost/dynamodb/starter.js:45:19)
    at Object.start (/Users/alexraducu/work/serverless/ssr-react-next/node_modules/dynamodb-localhost/index.js:20:32)
    at ServerlessDynamodbLocal.startHandler (/Users/alexraducu/work/serverless/ssr-react-next/node_modules/serverless-dynamodb-local/index.js:226:25)
    at BbPromise.reduce (/Users/alexraducu/.node/lib/node_modules/serverless/lib/classes/PluginManager.js:400:55)
From previous event:
    at PluginManager.invoke (/Users/alexraducu/.node/lib/node_modules/serverless/lib/classes/PluginManager.js:400:22)
    at PluginManager.run (/Users/alexraducu/.node/lib/node_modules/serverless/lib/classes/PluginManager.js:431:17)
    at variables.populateService.then.then (/Users/alexraducu/.node/lib/node_modules/serverless/lib/Serverless.js:114:33)
    at processImmediate (timers.js:632:19)
    at process.topLevelDomainCallback (domain.js:120:23)
From previous event:
    at Serverless.run (/Users/alexraducu/.node/lib/node_modules/serverless/lib/Serverless.js:101:6)
    at serverless.init.then (/Users/alexraducu/.node/lib/node_modules/serverless/bin/serverless:43:28)
    at /Users/alexraducu/.node/lib/node_modules/serverless/node_modules/graceful-fs/graceful-fs.js:111:16
    at /Users/alexraducu/.node/lib/node_modules/serverless/node_modules/graceful-fs/graceful-fs.js:45:10
    at FSReqCallback.args [as oncomplete] (fs.js:147:20)
From previous event:
    at initializeErrorReporter.then (/Users/alexraducu/.node/lib/node_modules/serverless/bin/serverless:43:6)
    at processImmediate (timers.js:632:19)
    at process.topLevelDomainCallback (domain.js:120:23)
From previous event:
    at /Users/alexraducu/.node/lib/node_modules/serverless/bin/serverless:28:46
    at Object.<anonymous> (/Users/alexraducu/.node/lib/node_modules/serverless/bin/serverless:67:4)
    at Module._compile (internal/modules/cjs/loader.js:721:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:732:10)
    at Module.load (internal/modules/cjs/loader.js:620:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:560:12)
    at Function.Module._load (internal/modules/cjs/loader.js:552:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:774:12)
    at executeUserCode (internal/bootstrap/node.js:342:17)
    at startExecution (internal/bootstrap/node.js:276:5)
    at startup (internal/bootstrap/node.js:227:5)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:743:3)
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com
 
  Your Environment Information -----------------------------
     OS:                     darwin
     Node Version:           11.6.0
     Serverless Version:     1.37.1`

**Comment by Buri** [comment]
Exact same issue:
```
 Your Environment Information -----------------------------
     OS:                     linux
     Node Version:           10.14.2
     Serverless Version:     1.37.1
```

**Comment by Alex-RR** [comment]
Anyone could help us please?

**Comment by keesvanlierop** [comment]
Reverting to 0.2.30 has for me fixed the issue

**Comment by Alex-RR** [comment]
I confirm, 0.2.30 works for me too. Thanks a lot!

**Comment by braunsquared** [comment]
Looks like the installHandler for the plugin isn't passing the proper options to dynamodbLocal.install.  The installHandler is passing (callback, path) when dynamodbLocal.install is expecting (config, callback)

My assumption is a version mismatch.

**Comment by braunsquared** [comment]
Reverting back to 0.2.35 did the job for me

**Comment by alexdilley** [comment]
I could install `0.2.36` with its *correct* set of dependencies only by installing `5ff42c28` specifically.

**Comment by samp64** [comment]
Getting the same issue. None of the versions suggested fix for me.

```
  Your Environment Information -----------------------------
     OS:                     darwin
     Node Version:           11.6.0
     Serverless Version:     1.37.1
```

Any suggestions would be appreciated!

**Comment by iamjoshua** [comment]
0.2.35 worked for me only _after_ removing the latest version and then running through the complete setup again: 

```
yarn remove serverless-dynamodb-local
yarn add serverless-dynamodb-local@0.2.35
sls dynamodb install
```

Also perhaps worth noting that if the java sdk isn't installed, an error will occur with a link just to java and not the sdk and so installing it will not help. Google java sdk for latest versions/OS.

**Comment by travjam2018** [comment]
I confirm above worked... used npm ...

**Comment by barroyo** [comment]
same here, using @0.2.35 works for me

**Comment by barroyo** [comment]
In case you want to use the latest version (0.2.36) check this issue https://github.com/99xt/serverless-dynamodb-local/issues/195. 

after upgrading running the install command like this fixed the issue for me (on MacOS)
`sls dynamodb install --localPath ./bin`

**Comment by ikitozen** [comment]
confirmed, downgrading to @0.2.35 worked

**Comment by kakchan** [comment]
`dynamodb/utils.js:absPath()` in the `dynamodb-localhost@0.0.5` package caused this issue

"0.2.36" is using older version `0.0.5` of `dynamodb-localhost`, whereas "0.2.35" is using`dynamodb-localhost@0.0.6`

the workaround would be update your `package.json`
```
  "resolutions": {
    "dynamodb-localhost": "0.0.7"   // or "0.0.6"
  }
```

**Comment by Dsantacruz** [comment]
this worked for me:

is exist error with version 0.2.36 serverless-dynamodb-local, then install:

Steps:

1) npm install serverless-dynamodb-local@0.2.35

2) sls dynamodb install 

3) sls offline start

**Comment by bearsworth** [comment]
If anyone is reading this, its easier to revert to 0.2.35 as the previous comment mentions. If you want to fix it though for 0.2.36 (temporarily) the issue is found in the installer.js. The first conditional hits and throws an error because the response from AWS servers is 200 and not 302. For some reason the developer wrote if (302 != response...) as the first condition. I'm guessing before the link would redirect somewhere, but now its no longer the case with the current AWS dynamodb link. So it will always hit that condition and throw an error.

**Comment by pveller** [comment]
It was something else on my side. The `utils.absPath()` in `dynamodb-localhost` always does this:
```javascript
return path.dirname(__filename) + "/" + relPath
```

When I launch `sls dynamodb install` from the VS code terminal, the `path` passed down is already an absolute path. The `install` command then just silently does nothing. 

I fixed the `absPath()` to be a little more intelligent but it's already fixed in the `dynamodb-localhost` as of 0.0.6. The latest version of `serverless-dynamodb-local` seems to install 0.0.5. I forced the latest using [yarn's resolutions](https://yarnpkg.com/lang/en/docs/selective-version-resolutions/) and it now installs and starts just fine:

```javascript
  "resolutions": {
    "serverless-dynamodb-local/dynamodb-localhost": "0.0.7"
  }
```

**Comment by bearsworth** [comment]
Ah, my issue was stemming from being unable to download. I read part of the comments here and was able to get it work on an older version.

**Comment by zackdotcomputer** [comment]
Should we take the fact that this bug has been live for over a month as an indicator that this project is abandonware now?

**Comment by jeremiahlukus** [comment]
Or you could submit a pull request

**Comment by zackdotcomputer** [comment]
@jeremiahlukus that would have been tough since I had no idea what the problem was beyond "the new version is broken" and it feels like it would've been bad form for a PR to just say "delete everything since the last version" 😉

Luckily for both of us @shalithasuranga already has submitted a pull request, #215, which I think aims to fix this issue via the solution found in issue #195 . Now we'll just wait and see if it gets merged in and released to NPM...

**Comment by shalithasuranga** [comment]
@zackzachariah #215 will be merged and npm module will be updated asap 👍

**Comment by shalithasuranga** [comment]
Please try new version `0.2.37`. The issue was fixed

**Comment by sethmoore** [comment]
I just got this same error on 0.2.37

**Comment by zackdotcomputer** [comment]
Have confirmed - the plugin worked for me on `0.2.37` (with the caveat that I ran `sls dynamodb remove` **before** upgrading away from `0.2.35`).

**Comment by ghost** [comment]
I fixed the problem with installing openjdk. I missed reading plugin requirements.  :-/

**Comment by vafl-brut** [comment]
Still having this issue :/ Tried everything in the thread - nothing seems to work :feelinghelpless:

**Comment by uccmen** [comment]
> Have confirmed - the plugin worked for me on `0.2.37` (with the caveat that I ran `sls dynamodb remove` **before** upgrading away from `0.2.35`).

@zackzachariah, @shalithasuranga  it didn't work for me. here's what my yarn.lock file looks like
```
serverless-dynamodb-local@^0.2.37:
  version "0.2.37"
  resolved "https://registry.yarnpkg.com/serverless-dynamodb-local/-/serverless-dynamodb-local-0.2.37.tgz#cbd7225c8d45e3b28a7644ec2845e31c77a13a6a"
  integrity sha512-1q3rfmn+Y1nZQhmuVdFLJcMTc6la1zY6sT98TS6Af0vRG//oa7CuiDEox0XEzuj7KZ0TodwmXmnRdzt9VRm3fA==
  dependencies:
    aws-sdk "^2.7.0"
    bluebird "^3.4.6"
    dynamodb-localhost "^0.0.7"
    lodash "^4.17.0"
```

What worked for me is when I locked `dynamodb-localhost` at `0.0.5`

```
"resolutions": {
    "dynamodb-localhost": "0.0.5"
  }
```

**Comment by dimkomkd420** [comment]
Guys, I've tried all solutions but didn't work for me, what works is installing JRE (Java Runtime Engine). 
DynamoDB local requires Java Runtime Engine version 6.x or newer. Make sure u have installed JRE (Java Runtime Engine) on your machine.

**Comment by uccmen** [comment]
@dimkomkd420 what's your JRE version?

**Comment by dimkomkd420** [comment]
@uccmen 
java version "1.8.0_211"
Java(TM) SE Runtime Environment (build 1.8.0_211-b12)
Java HotSpot(TM) Client VM (build 25.211-b12, mixed mode, sharing)

**Comment by superjose** [comment]
TL;DR;

1.  Check which process is using port 8000:

- (Windows)
`Get-Process -Id (Get-NetTCPConnection -LocalPort 8000).OwningProcess`
- (MacOS)
`netstat -vanp tcp | grep 8000`
- (Ubuntu) (Have [netstat installed](https://vitux.com/find-out-which-port-number-a-process-is-listening-on-using-linux/))
`sudo netstat -ltnp | grep -w ':8000'`
2.  Kill the process

3. Run serverless once more.(Adapt the following to your needs):
`serverless dynamodb install && serverless offline start --watch`

<hr/>

Ok. This took a while, but this nailed it. Hopefully this will be able to solve your issue as well. I'm currently using Windows 10 May 2019 update (Not that it matters, I think...)

The problem (in my case) was that another application was using the port 8000 which dynamodb uses. Therefore, there was a conflict and the process couldn't start. 

A great place to start debugging is by manually running the dynamodb .jar file:
From the [Amazon Docs](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBLocal.DownloadingAndRunning.html) If you go where the DynamoDb is installed (In my case was next to node_modules), just run :
`java -D"java.library.path=./DynamoDBLocal_lib" -jar DynamoDBLocal.jar`

This will give you further information. In my case, the port 8000 was occupied by another process. 

I then used PowerShell to check which process was using it:
`Get-Process -Id (Get-NetTCPConnection -LocalPort 8000).OwningProcess`

I looked for the process and killed it.

Then I ran (Adapt the following to your needs):
`serverless dynamodb install && serverless offline start --watch`


Note: 
You can always download DynamoDB (in case you can't find it from serverless) and run it anywhere from your command line. 

You will also need [Java Runtime Environment - JRE](https://filehippo.com/download_jre_64/) (You don't need the SDK to run it)

**Comment by Ragers** [comment]
> @uccmen
> java version "1.8.0_211"
> Java(TM) SE Runtime Environment (build 1.8.0_211-b12)
> Java HotSpot(TM) Client VM (build 25.211-b12, mixed mode, sharing)

look at your java version. Looking at the documentation you require java version 6.x or higher. :)
see here: https://serverless.com/plugins/serverless-dynamodb-local/

**Comment by ezweave** [comment]
@Ragers Java has weird version names.  `1.8.*` is "Java 8.x", so that is the correct version.

Per [Oracle](https://www.oracle.com/java/technologies/javase/jdk8-naming.html):
> In JDK 8 and JRE 8, the version strings are 1.8 and 1.8.0.

There is no such thing as a Java internal version number "6.x" yet.  It is `1.6.*`.

**Comment by middoors** [comment]
don't forget to run "sls dynamodb install" before running "sls offline start"
